### PR TITLE
fix: convert nodelist to array before passing to native Array methods

### PR DESCRIPTION
**Pull request by ravijayaramappa**
## Details
Alternative to: https://github.com/salesforce/lwc/pull/1546

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

If yes, please describe the impact and migration path for existing applications.

## The PR fulfills these requirements:
* Have tests for the proposed changes been added? ✅ 
* Have you followed [these instructions](../CONTRIBUTING.md#-commit-message-conventions) to clearly describe the issue being fixed or feature enhanced? ✅

### CODE REVIEW

File: packages/@lwc/synthetic-shadow/src/shared/node-collection-util.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * Issue #1545 - Writing a custom util to convert NodeList and HTMLCollection instances to array.
+ * This is to aviod conflict with some legacy third party libraries like prototype.js
+ * that patch Array.prototype and are incapable of handling NodeList and HTMLCollection.
+ */
+export function arrayFromCollection<T extends Node, K extends Element>(
+    collection: NodeListOf<T> | HTMLCollectionOf<K>
+): Array<T | K> {
+    const size = collection.length;
+    const cloned: T[] | K[] = new ArrayConstructor(size);
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/shared/node-collection-util.ts]: This method performs better than instantiating an empty array and calling ArrayPush.
https://jsperf.com/new-array-vs-push-lwc
Screen Shot 2019-10-01 at 1 51 29 PM

File: packages/@lwc/synthetic-shadow/src/shared/node-collection-util.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { ArrayConstructor } from '@lwc/shared';
+/**
+ * Issue #1545 - Writing a custom util to convert NodeList and HTMLCollection instances to array.
+ * This is to aviod conflict with some legacy third party libraries like prototype.js
+ * that patch Array.prototype and are incapable of handling NodeList and HTMLCollection.
+ */
+export function arrayFromCollection<T extends Node, K extends Element>(
```
Comment (caridy) [inline packages/@lwc/synthetic-shadow/src/shared/node-collection-util.ts]: can we have two methods, one for htmlcollection and one for nodelist? even if the underlying implementation is the same... I feel that this PR is messing with the types.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/traverse.ts [hunk 3]
```diff
@@ -83,7 +84,7 @@ export function getAllMatches(owner: Element, nodeList: NodeList | Node[]): Arra
     return filteredAndPatched;
 }
 
-export function getFirstMatch(owner: Element, nodeList: NodeList): Element | null {
+export function getFirstMatch(owner: Element, nodeList: NodeList | Element[]): Element | null {
```
Comment (caridy) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/traverse.ts]: this type extension seems wrong.

Comment (ravijayaramappa) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/traverse.ts]: This method is called in two places. One with a NodeList and the other with an Element[]. That is the reason for this change.
BTW, the same pattern is in L73 
`export function getAllMatches(owner: Element, nodeList: NodeList | Node[]):`

File: packages/@lwc/synthetic-shadow/src/faux-shadow/element.ts [hunk 4]
```diff
@@ -332,7 +333,9 @@ if (hasOwnProperty.call(HTMLElement.prototype, 'children')) {
 // Deep-traversing patches from this point on:
 
 function querySelectorPatched(this: Element /*, selector: string*/): Element | null {
-    const nodeList = elementQuerySelectorAll.apply(this, ArraySlice.call(arguments) as [string]);
+    const nodeList = arrayFromCollection(
+        elementQuerySelectorAll.apply(this, ArraySlice.call(arguments) as [string])
```
[the remark this answers is not in the document]
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/element.ts]: @ekashida the problem with this change is that if the user makes a mistake, and calls `elm.querySelector()`, no error will be issued, while in the existing code, the error will be throw by the host:

> `Failed to execute 'querySelector' on 'Document': 1 argument required, but only 0 present.`

[the remark this answers is not in the document]
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/element.ts]: basically, when patching original methods, we need to preserve the same errors, always propagating the original argument list, and that's only possible if using `apply`

**Pull request merged**